packages/flutter/lib/src/widgets/focus_traversal.dart

```diff
@@ -0,0 +1,783 @@
+///     focus traversal in a direction.
+abstract class FocusTraversalPolicy {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const FocusTraversalPolicy();
+
+  /// Returns the node that should receive focus if there is no current focus
+  /// in the scope to which the [currentNode] belongs.
+  ///
+  /// This is used by [next]/[previous]/[inDirection] to determine which node to
+  /// focus if they are called, but no node is currently focused.
+  ///
+  /// It is also used by the [FocusManager] to know which node to focus
+  /// initially if no nodes are focused.
+  ///
+  /// If the [direction] is null, then it should find the appropriate first node
+  /// for next/previous, and if direction is non-null, should find the
+  /// appropriate first node in that direction.
+  ///
+  /// The [currentNode] argument must not be null.
+  FocusNode findFirstFocus(FocusNode currentNode);
+
+  /// Returns the node in the given [direction] that should receive focus if
+  /// there is no current focus in the scope to which the [currentNode] belongs.
+  ///
+  /// This is typically used by [inDirection] to determine which node to focus
+  /// if it is called, but no node is currently focused.
+  ///
+  /// All arguments must not be null.
+  FocusNode findFirstFocusInDirection(FocusNode currentNode, TraversalDirection direction);
+
+  /// Focuses the next widget in the focus scope that contains the given
+  /// [currentNode].
+  ///
+  /// This should determine what the next node to receive focus should be by
+  /// inspecting the node tree, and then calling [FocusNode.requestFocus] on
+  /// the node that has been selected.
+  ///
+  /// Returns true if it successfully found a node and requested focus.
+  ///
+  /// The [currentNode] argument must not be null.
+  bool next(FocusNode currentNode);
+
+  /// Focuses the previous widget in the focus scope that contains the given
+  /// [currentNode].
+  ///
+  /// This should determine what the previous node to receive focus should be by
+  /// inspecting the node tree, and then calling [FocusNode.requestFocus] on
+  /// the node that has been selected.
+  ///
+  /// Returns true if it successfully found a node and requested focus.
+  ///
+  /// The [currentNode] argument must not be null.
+  bool previous(FocusNode currentNode);
+
+  /// Focuses the next widget in the given [direction] in the focus scope that
+  /// contains the given [currentNode].
+  ///
+  /// This should determine what the next node to receive focus in the given
+  /// [direction] should be by inspecting the node tree, and then calling
+  /// [FocusNode.requestFocus] on the node that has been selected.
+  ///
+  /// Returns true if it successfully found a node and requested focus.
+  ///
+  /// If this function returns true when called by the subclass, then the
```

Comment:
Wouldn't this be true for all methods? I assume what we're concerned about is if a subclass calls super.next/previous/indirection() and it returns true then that subclass should call requestFocus on something else.

Reply:
It shouldn't call `requestFocus` on anything else because the parent class implementation will have done it already.  I suppose this should be self-evident, so I'm going to just remove this comment.